pkg/resources/external_function.go

```diff
@@ -196,216 +193,196 @@ func ExternalFunction() *schema.Resource {
 		for _, header := range d.Get("header").(*schema.Set).List() {
-			headerDef := map[string]string{}
-			for key, val := range header.(map[string]interface{}) {
-				headerDef[key] = val.(string)
-			}
-			headers = append(headers, headerDef)
+			m := header.(map[string]interface{})
+			headerName := m["name"].(string)
+			headerValue := m["value"].(string)
+			headers = append(headers, sdk.ExternalFunctionHeaderRequest{
+				Name:  headerName,
+				Value: headerValue,
+			})
 		}
-
-		builder.WithHeaders(headers)
+		req.WithHeaders(headers)
 	}
 
 	if v, ok := d.GetOk("context_headers"); ok {
-		contextHeaders := expandStringList(v.([]interface{}))
-		builder.WithContextHeaders(contextHeaders)
+		contextHeadersList := expandStringList(v.([]interface{}))
+		contextHeaders := make([]sdk.ExternalFunctionContextHeaderRequest, 0)
+		for _, header := range contextHeadersList {
+			contextHeaders = append(contextHeaders, sdk.ExternalFunctionContextHeaderRequest{
+				ContextFunction: header,
+			})
+		}
+		req.WithContextHeaders(contextHeaders)
 	}
 
 	if v, ok := d.GetOk("max_batch_rows"); ok {
-		builder.WithMaxBatchRows(v.(int))
+		req.WithMaxBatchRows(sdk.Int(v.(int)))
 	}
 
 	if v, ok := d.GetOk("compression"); ok {
-		builder.WithCompression(v.(string))
+		req.WithCompression(sdk.String(v.(string)))
 	}
+
 	if v, ok := d.GetOk("request_translator"); ok {
-		builder.WithRequestTranslator(v.(string))
+		req.WithRequestTranslator(sdk.Pointer(sdk.NewSchemaObjectIdentifierFromFullyQualifiedName(v.(string))))
 	}
+
 	if v, ok := d.GetOk("response_translator"); ok {
-		builder.WithResponseTranslator(v.(string))
+		req.WithResponseTranslator(sdk.Pointer(sdk.NewSchemaObjectIdentifierFromFullyQualifiedName(v.(string))))
 	}
 
-	stmt := builder.Create()
-	if err := snowflake.Exec(db, stmt); err != nil {
-		return fmt.Errorf("error creating external function %v err = %w", name, err)
+	if err := client.ExternalFunctions.Create(ctx, req); err != nil {
+		return diag.FromErr(err)
 	}
 
-	externalFunctionID := &externalFunctionID{
-		DatabaseName:             database,
-		SchemaName:               dbSchema,
-		ExternalFunctionName:     name,
-		ExternalFunctionArgTypes: argtypes,
-	}
-	dataIDInput, err := externalFunctionID.String()
-	if err != nil {
-		return err
-	}
-	d.SetId(dataIDInput)
+	snowflakeID := helpers.EncodeSnowflakeID(databaseName, schemaName, name, argtypes)
+	d.SetId(snowflakeID)
 
-	return ReadExternalFunction(d, meta)
+	return ReadContextExternalFunction(ctx, d, meta)
 }
 
-// ReadExternalFunction implements schema.ReadFunc.
-func ReadExternalFunction(d *schema.ResourceData, meta interface{}) error {
+func ReadContextExternalFunction(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	db := meta.(*sql.DB)
-	externalFunctionID, err := externalFunctionIDFromString(d.Id())
-	if err != nil {
-		return err
+	client := sdk.NewClientFromDB(db)
+	parts := strings.Split(d.Id(), helpers.IDDelimiter)
```

Comment:
why not `DecodeSnowflakeID()`?

Reply:
n/a anymore, there was no need to encode it in the first place. using a simple `sdk.NewSchemaObjectIdentifierFromFullyQualifiedName(d.Id())` instead